ext/drx/drx_buf.c

```diff
@@ -648,6 +649,96 @@ drx_buf_insert_buf_store(void *drcontext, drx_buf_t *buf, instrlist_t *ilist,
     }
 }
 
+static void
+insert_load(void *drcontext, instrlist_t *ilist, instr_t *where,
+                    reg_id_t dst, reg_id_t src, opnd_size_t opsz)
+{
+    switch (opsz) {
+    case OPSZ_1:
+        MINSERT(ilist, where, XINST_CREATE_load_1byte
+                (drcontext,
+                 opnd_create_reg(reg_resize_to_opsz(dst, opsz)),
+                 opnd_create_base_disp(src, DR_REG_NULL, 0, 0, opsz)));
+        break;
+    case OPSZ_2:
+        MINSERT(ilist, where, XINST_CREATE_load_2bytes
+                (drcontext,
+                 opnd_create_reg(reg_resize_to_opsz(dst, opsz)),
+                 opnd_create_base_disp(src, DR_REG_NULL, 0, 0, opsz)));
+        break;
+    case OPSZ_4:
+#if defined(X86_64) || defined(AARCH64)
+    case OPSZ_8:
+#endif
+        MINSERT(ilist, where, XINST_CREATE_load
+                (drcontext,
+                 opnd_create_reg(reg_resize_to_opsz(dst, opsz)),
+                 opnd_create_base_disp(src, DR_REG_NULL, 0, 0, opsz)));
+        break;
+    default:
+        DR_ASSERT(false);
+        break;
+    }
+}
+
+DR_EXPORT
+void
+drx_buf_insert_buf_memcpy(void *drcontext, drx_buf_t *buf, instrlist_t *ilist,
+                          instr_t *where, reg_id_t dst, reg_id_t src, ushort len)
+{
+    opnd_size_t opsz = opnd_size_from_bytes(len);
+
+    if (len > sizeof(app_pc)) {
+        /* slow path, we call into our personal memcpy() implementation */
+        /* XXX: We should elect to write our own memcpy() built in DR IR. This
+         * would potentially be slower than the glibc memcpy, but then we can
+         * incrementally update the buffer pointer so the user can recover
+         * from partial writes.
+         * XXX: It's also unlikely that if glibc's memcpy faults, we'd be able
+         * to recover from it -- an in-app fault will not be handled by us, and
+         * also the buffer pointer is not updated incrementally.
```

Comment:
It may take some work to handle faults in the memcpy, yes.  But until we either handle the faults, or explicitly check for the end of the buffer, isn't this not yet ready for users to use?  Won't someone using this risk crashing if the memcpy goes off the end and hits the ro page?  So it should have a FIXME here and a not-fully-finished note in the header (and commit msg) or sthg so we know it's not yet complete?